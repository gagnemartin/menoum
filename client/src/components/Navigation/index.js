import { useEffect, useRef, useState } from 'react'
import { Link as RouterLink } from 'react-router-dom'
import Box from '@mui/material/Box'
import Link from '@mui/material/Link'
import MenuList from '@mui/material/MenuList'
import MenuItem from '@mui/material/MenuItem'
import Popper from '@mui/material/Popper'
import Paper from '@mui/material/Paper'
import ClickAwayListener from '@mui/material/ClickAwayListener'
import Grow from '@mui/material/Grow'
import AccountCircleIcon from '@mui/icons-material/AccountCircle'
import { styled } from '@mui/material/styles'
import { useUserState } from '../../hooks/useUser'
import Logout from '../Logout'

const NavLink = styled(Link)(({ theme }) => ({
  padding: theme.spacing(2)
}))

const Navigation = () => {
  const userState = useUserState()
  const { loading, user } = userState

  const [open, setOpen] = useState(false)
  const anchorRef = useRef(null)

  const handleToggle = () => {
    setOpen((prevOpen) => !prevOpen)
  }

  const handleClose = (event) => {
    if (anchorRef.current && anchorRef.current.contains(event.target)) {
      return
    }

    setOpen(false)
  }

  function handleListKeyDown(event) {
    if (event.key === 'Tab') {
      event.preventDefault()
      setOpen(false)
    }
  }
  // return focus to the button when we transitioned from !open -> open
  const prevOpen = useRef(open)
  useEffect(() => {
    if (prevOpen.current === true && open === false) {
      anchorRef?.current?.focus()
    }

    prevOpen.current = open
  }, [open])

  return (
    <nav>
      <Box display='flex' alignItems='center' px={2}>
        <Box display='flex' flexGrow={1}>
          <NavLink component={RouterLink} to='/' data-testid='nav-home'>
            Menoum
          </NavLink>
        </Box>
        <Box display='flex'>
          {!loading ? (
            user?.email ? (
              <>
                <NavLink
                  ref={anchorRef}
                  component='button'
                  aria-controls='nav-main-menu'
                  aria-haspopup='true'
                  onClick={handleToggle}
                  data-testid='nav-email'
                >
                  <AccountCircleIcon color='disabled' sx={{ '&:hover': { color: 'black' } }} />
                </NavLink>

                <Popper open={open} anchorEl={anchorRef.current} role={undefined} transition disablePortal>
                  {({ TransitionProps, placement }) => (
                    <Grow {...TransitionProps} style={{ transformOrigin: placement === 'bottom' ? 'center top' : 'center bottom' }}>
                      <Paper>
                        <ClickAwayListener onClickAway={handleClose}>
                          <MenuList autoFocusItem={open} id='nav-main-menu' onKeyDown={handleListKeyDown}>
                            {user.role === 'admin' && (
                              <MenuItem onClick={handleClose}>
                                <Link component={RouterLink} to='/recipe/new' data-testid='nav-new-recipe'>
                                  Add a Recipe
                                </Link>
                              </MenuItem>
                            )}
                            <MenuItem onClick={handleClose}>
                              <Logout />
                            </MenuItem>
                          </MenuList>
                        </ClickAwayListener>
                      </Paper>
                    </Grow>
                  )}
                </Popper>
              </>
            ) : (
              <>
                <NavLink component={RouterLink} to='/login' data-testid='nav-login'>
                  Sign in
                </NavLink>
                <NavLink component={RouterLink} to='/register' data-testid='nav-register'>
                  Register
                </NavLink>
              </>
            )
          ) : (
            <p data-testid='nav-loading'>loading</p>
          )}
        </Box>
      </Box>
    </nav>
  )
}

export default Navigation
