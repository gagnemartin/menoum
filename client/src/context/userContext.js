import { createContext, useEffect, useReducer } from 'react'
import { UsersService } from '../services'
import { isSuccessResponse } from '../global/helpers'
import userReducer, { actionTypes, initialState } from '../reducers/userReducer'

const UserStateContext = createContext(initialState)
const UserDispatchContext = createContext({})

const refresh = async (dispatch) => {
  dispatch({ type: actionTypes.request, loading: true })
  try {
    const payload = await UsersService.refresh()

    if (isSuccessResponse(payload)) {
      dispatch({ type: actionTypes.success, payload })
    } else {
      dispatch({ type: actionTypes.error, error: payload })
    }

    return setInterval(async () => {
      const payload = await UsersService.refresh()

      if (isSuccessResponse(payload)) {
        dispatch({ type: actionTypes.success, payload })
      } else {
        dispatch({ type: actionTypes.error, error: payload })
      }
    }, 840000)
  } catch (error) {
    dispatch({ type: actionTypes.error, error })
  }
}

const login = async (dispatch, data) => {
  dispatch({ type: actionTypes.request, loading: true })
  try {
    const payload = await UsersService.login(data)

    if (isSuccessResponse(payload)) {
      dispatch({ type: actionTypes.success, payload })
      return payload
    } else {
      throw payload
    }
  } catch (error) {
    dispatch({ type: actionTypes.error, error })
    return error
  }
}

const register = async (dispatch, data) => {
  dispatch({ type: actionTypes.request, loading: true })
  try {
    const payload = await UsersService.register(data)

    if (isSuccessResponse(payload)) {
      dispatch({ type: actionTypes.success, payload })
      return payload
    } else {
      throw payload
    }
  } catch (error) {
    dispatch({ type: actionTypes.error, error })
    return error
  }
}

const logout = async (dispatch) => {
  dispatch({ type: actionTypes.request, loading: true })
  try {
    const payload = await UsersService.logout()

    if (isSuccessResponse(payload)) {
      console.log('SUCCESS', payload)
      dispatch({ type: actionTypes.success, payload })
      return payload
    } else {
      throw payload
    }
  } catch (error) {
    dispatch({ type: actionTypes.error, error })
    return error
  }
}

const UserProvider = ({ children }) => {
  const [user, setUser] = useReducer(userReducer, initialState)

  useEffect(() => {
    const interval = refresh(setUser)

    return () => {
      clearInterval(interval)
    }
  }, [])

  return (
    <UserStateContext.Provider value={user}>
      <UserDispatchContext.Provider value={setUser}>{children}</UserDispatchContext.Provider>
    </UserStateContext.Provider>
  )
}

export { UserProvider, UserStateContext, UserDispatchContext, login, logout, register }
