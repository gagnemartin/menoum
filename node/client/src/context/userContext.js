import { createContext, useEffect, useContext, useReducer } from 'react'
import jwt from 'jsonwebtoken'
import { UsersService } from '../services'
import { isSuccessResponse } from '../global/helpers'

export const ACTION_TYPES = {
  request: 'REQUEST',
  success: 'SUCCESS',
  error: 'ERROR'
}

export const DEFAULT_STATE = {
  status: ACTION_TYPES.request,
  user: {},
  loading: true,
  error: {}
}

export const UserStateContext = createContext(DEFAULT_STATE)
export const UserDispatchContext = createContext({})

const userReducer = (state, action) => {
  switch (action.type) {
    case ACTION_TYPES.request: {
      return {
        ...DEFAULT_STATE,
        status: ACTION_TYPES.request,
        user: state.user,
        loading: true
      }
    }
    case ACTION_TYPES.success: {
      let user = {}

      if (action.payload.data && action.payload.data.token) {
        const {
          data: { token, expires_at }
        } = action.payload

        user = {
          ...jwt.decode(token),
          token,
          expires_at
        }
      }

      return {
        ...DEFAULT_STATE,
        status: ACTION_TYPES.success,
        loading: false,
        user
      }
    }
    case ACTION_TYPES.error: {
      return {
        ...DEFAULT_STATE,
        status: ACTION_TYPES.error,
        loading: false,
        error: action.error
      }
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}

const useUserState = () => {
  const context = useContext(UserStateContext)
  if (context === undefined) {
    throw new Error('useUserState must be used within a UserProvider')
  }
  return context
}

const useUserDispatch = () => {
  const context = useContext(UserDispatchContext)
  if (context === undefined) {
    throw new Error('useUserDispatch must be used within a UserProvider')
  }
  return context
}

const refresh = async (dispatch) => {
  dispatch({ type: ACTION_TYPES.request, loading: true })
  try {
    const payload = await UsersService.refresh()

    if (isSuccessResponse(payload)) {
      dispatch({ type: ACTION_TYPES.success, payload })
    } else {
      dispatch({ type: ACTION_TYPES.error, error: payload })
    }

    return setInterval(async () => {
      const payload = await UsersService.refresh()

      if (isSuccessResponse(payload)) {
        dispatch({ type: ACTION_TYPES.success, payload })
      } else {
        dispatch({ type: ACTION_TYPES.error, error: payload })
      }
    }, 840000)
  } catch (error) {
    dispatch({ type: ACTION_TYPES.error, error })
  }
}

const login = async (dispatch, data) => {
  dispatch({ type: ACTION_TYPES.request, loading: true })
  try {
    const payload = await UsersService.login(data)

    if (isSuccessResponse(payload)) {
      dispatch({ type: ACTION_TYPES.success, payload })
      return payload
    } else {
      throw payload
    }
  } catch (error) {
    dispatch({ type: ACTION_TYPES.error, error })
    return error
  }
}

const register = async (dispatch, data) => {
  dispatch({ type: ACTION_TYPES.request, loading: true })
  try {
    const payload = await UsersService.register(data)

    if (isSuccessResponse(payload)) {
      dispatch({ type: ACTION_TYPES.success, payload })
      return payload
    } else {
      throw payload
    }
  } catch (error) {
    dispatch({ type: ACTION_TYPES.error, error })
    return error
  }
}

const logout = async (dispatch) => {
  dispatch({ type: ACTION_TYPES.request, loading: true })
  try {
    const payload = await UsersService.logout()

    if (isSuccessResponse(payload)) {
      dispatch({ type: ACTION_TYPES.success, payload })
      return payload
    } else {
      throw payload
    }
  } catch (error) {
    dispatch({ type: ACTION_TYPES.error, error })
    return error
  }
}

const useUser = () => {
  return [useUserState(), useUserDispatch()]
}

const UserProvider = ({ children }) => {
  const [user, setUser] = useReducer(userReducer, DEFAULT_STATE)

  useEffect(() => {
    const interval = refresh(setUser)

    return () => {
      clearInterval(interval)
    }
  }, [])

  return (
    <UserStateContext.Provider value={user}>
      <UserDispatchContext.Provider value={setUser}>{children}</UserDispatchContext.Provider>
    </UserStateContext.Provider>
  )
}

export { UserProvider, useUserState, useUserDispatch, userReducer, useUser, login, logout, register }
