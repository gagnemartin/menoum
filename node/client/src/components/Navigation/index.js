import { useState, useEffect } from 'react';
import { Link } from 'react-router-dom'
import { useUserState } from '../../context/userContext'
import Logout from '../Logout'

const Navigation = () => {
  const userState = useUserState()
  const [isLoading, setIsLoading] = useState(true)

  useEffect(() => {
    if (userState.status !== 'REQUEST') {
      setIsLoading(false)
    }
  }, [userState])

  return (
    <nav>
      <ul>
        <li>
          <Link to='/' data-testid='nav-home'>
            Home
          </Link>
        </li>

        {!isLoading && Object.keys(userState.user).length === 0 ? (
          <>
            <li>
              <Link to='/login' data-testid='nav-login'>
                Login
              </Link>
            </li>
            <li>
              <Link to='/register' data-testid='nav-register'>
                Register
              </Link>
            </li>
          </>
        ) : (
          <>
            {userState.user.role === 'admin' && (
              <li>
                <Link to='/recipe/new' data-testid='nav-new-recipe'>
                  Add a Recipe
                </Link>
              </li>
            )}
            <li data-testid='nav-email'>{userState.user.email}</li>
            <li>
              <Logout />
            </li>
          </>
        )}
      </ul>
    </nav>
  )
}

export default Navigation
