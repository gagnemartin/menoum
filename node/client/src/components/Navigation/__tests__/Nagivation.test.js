import { useReducer } from 'react'
import { render } from '@testing-library/react'
import { renderHook, act } from '@testing-library/react-hooks'
import { BrowserRouter as Router } from 'react-router-dom'
import Navigation from '../'
import { UserProvider, DEFAULT_STATE, ACTION_TYPES, userReducer } from '../../../context/userContext'

/**
 * A custom render to setup providers. Extends regular
 * render options with `providerProps` to allow injecting
 * different scenarios to test with.
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
const customRender = (children) => {
  return render(<UserProvider>{children}</UserProvider>)
}

describe('<Navigation />', () => {
  it('should not have login button when loading', () => {
    const updatedState = {
      ...DEFAULT_STATE,
      status: ACTION_TYPES.success,
      user: {
        email: 'test@test.com'
      },
      loading: false
    }
    const { result, rerender } = renderHook(() => useReducer(userReducer, DEFAULT_STATE))
    const [state, dispatch] = result.current
    // dispatch({ type: ACTION_TYPES.success, data: { token: 'as' } })
    act(() => {
      dispatch({
        type: ACTION_TYPES.success,
        payload: {
          data: {
            token:
              'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6InRlc3RAdGVzdC5jb20iLCJuYW1lIjoiSm9obiBEb2UiLCJpYXQiOjE1MTYyMzkwMjJ9.J78feKgEv-EbpTKGLYT74KT-qtaXjZDAGlpd3Rxwhy4'
          }
        }
      })
      // rerender({
      //   status: ACTION_TYPES.success,
      //   user: {
      //     email: 'test@test.com'
      //   },
      //   loading: false
      // })
    })
    rerender(updatedState)

    const { queryByTestId, queryByText, asFragment } = customRender(
      <Router>
        <Navigation />
      </Router>
    )

    // customRender(<Navigation />, { providerProps })
    //const { getByTestId, asFragment } = render(<Navigation />)
    const element = queryByText(updatedState.user.email)

    expect(element).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })
})
