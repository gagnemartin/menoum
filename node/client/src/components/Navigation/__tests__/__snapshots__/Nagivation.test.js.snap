// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> should not have login button when loading 1`] = `
<DocumentFragment>
  <nav>
    <ul>
      <li>
        <a
          data-testid="nav-home"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        data-testid="nav-email"
      />
      <li>
        <a
          data-testid="nav-logout"
          href="#"
        >
          Logout
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
